### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by angelortiz1007**
Preliminary IPC-Channel modifications based on rust 2018.  It is also a rebase of #166.

### CODE REVIEW

**Comment by jdm** [comment]
This work will require https://github.com/retep998/winapi-rs/pull/782.

File: src/platform/windows/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1851 @@
...
+}
+
+impl PartialEq for WinHandle {
+    fn eq(&self, other: &WinHandle) -> bool {
+        // WPre-Windows 10 API GetFileInformationByHandle() is used to
+        // to compare the underlying objects the handles refer to,
+        //
+        // On Windows 10, we could use:
+        // ```
+        // unsafe { winapi::um::handleapi::CompareObjectHandles(self.h, other.h) == winapi::shared::minwindef::TRUE }
```
Comment (Xanewok) [inline src/platform/windows/mod.rs]: It looks like we don't actually use `CompareObjectHandles` (all uses are commented out), so this shouldn't lead to link errors (as reported in https://github.com/retep998/winapi-rs/issues/781). Any idea what may be causing that?
@jdm @angelortiz1007 @retep998

Comment (Xanewok) [reply src/platform/windows/mod.rs]: I'm sorry for the noise, it's clearly defined in winapi-rs https://docs.rs/winapi/0.3.7/src/winapi/um/handleapi.rs.html#23-26, whereas I thought the snippets are for illustration purposes only. Will continue in the linked issue

**Comment by angelortiz1007** [comment]
No worries.

 

Let me know if I can help.

 

Regards,

Angel

**Comment by vlad20012** [comment]
What's the current status of this PR?

**Comment by jdm** [comment]
We need to:
* clean up the broken code in WinHandle's PartialEq impl
* add a cargo feature to enable using CompareObjectHandles
* link with WindowsApp.lib when that cargo feature is enabled

At that point I think this work will be ready to merge. If anybody would like to help with making that happen, we would appreciate it!

**Comment by retep998** [comment]
It might be better to dynamically load `CompareObjectHandles` using something like `libloading`.

**Comment by jdm** [comment]
Oh right, yes. That sounds like a the right way forward.

**Comment by ofek** [comment]
This is exciting!!!

**Comment by sunjay** [comment]
Thanks for everyone's work on this! Any update on this? `ipc-channel` seems to be one of the best crates for doing IPC in Rust right now. Would be great to get Windows support so it can be used cross-platform. :)

**Comment by antrik** [comment]
I guess I should really get back on this one of these days... My previous work was *almost* ready to merge (just missing a proper commit message for my latest change, and perhaps some additional test cases) -- but after I have been away for a long time, and new issues came up that I will need to dig into, the thought of getting back to it somewhat scares me :-(

The other thing is that with all the changes I have done to @vvuk 's original work, it probably needs another full review... Ideally, @vvuk himself, or someone else understanding this stuff, could look at all the commits aside from the one incorporating @vvuk 's original patch -- though at this point, maybe it would be more reasonable to just review the end result, with no regard for the history?...

**Comment by jdm** [comment]
https://github.com/servo/ipc-channel/pull/233#issuecomment-534597659 remains a summary of the work that is required here, to the best of my knowledge.

**Comment by BatmanAoD** [comment]
@antrik is your work-so-far in a public branch somewhere? Which of the issues in @jdm's comment does it address (fully or partially)?

**Comment by vvuk** [comment]
I think you just want to look at the end result.  The history doesn't really matter at this point.  But I'm probably not a good reviewer; I am very far from this at this point, and (sadly) not doing much Rust lately.

Also, the origin of this work is now 4 years old.  Was it really worth not landing the bulk of the work in an imperfect state, and then letting people make incremental improvements & fixes?  Hopefully this doesn't go another 4 years.

**Comment by Boscop** [comment]
Any update on this? :)

**Comment by jdm** [comment]
No update. https://github.com/servo/ipc-channel/pull/233#issuecomment-534597659 describes the work that remains for this branch.

**Comment by Boscop** [comment]
@jdm But aren't people using some branch of this already in production (so it's working on windows)? E.g. habitat.

**Comment by jdm** [comment]
I have no idea if anybody is using this branch. Servo is not.

**Comment by vvuk** [comment]
Does that mean that Servo is (still) not multiprocess on Windows?  Or is it doing an alternative approach instead of ipc-channel for communication?

**Comment by jdm** [comment]
Servo is still not multiprocess on Windows.

**Comment by vvuk** [comment]
That seems like a very poor outcome, because I had multiprocess servo working on Windows with a previous version of this ipc-channel work in October 2016, over 4 years ago.  I know Windows was never important for Servo (and still isn't), but I think this is doing a disservice to Windows users and developers.   Then again, the work described in comment #233 seems reasonable, but I'm _really_ hesitant to give it a shot because of how this went last time.  It was a pretty demoralizing experience.

**Comment by Boscop** [comment]
@vvuk Could you please try to make it work?
I would really appreciate to have this crate working on windows. (I would offer my help but this is outside of my expertise.) 
It would be a shame if all this effort didn't get merged. 
(I agree, that the last working state should have been merged, and further improvements could have been done in separate PRs.)

**Comment by jdm** [comment]
If someone finishes the work I'll merge it. I no longer have access to a Windows development environment.

**Comment by ofek** [comment]
I would still love to use this on Windows!

**Comment by jdm** [comment]
Fixed in #272

**Pull request closed without merging**